### PR TITLE
[llvm][ctx_profile] Add instrumentation lowering

### PR DESCRIPTION
**Pull request by mtrofin**
This adds the instrumentation lowering pass.

(Tracking Issue: #89287, RFC referenced there)

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfLowering.cpp [hunk 1]
```diff
@@ -22,3 +31,295 @@ static cl::list<std::string> ContextRoots(
...
+        return std::make_pair(NrCounters, NrCallsites);
+    }
+  }
+  return {0, 0};
+}
+} // namespace
+
+// set up tie-in with compiler-rt.
+// NOTE!!!
+// These have to match compiler-rt/lib/ctx_profile/CtxInstrProfiling.h
```
Comment (mtrofin) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfLowering.cpp:89]: Note: this is in PR #89838

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfLowering.cpp [hunk 2]
```diff
@@ -22,3 +32,298 @@ static cl::list<std::string> ContextRoots(
...
+      } else if (TheRootContext && isa<ReturnInst>(I)) {
+        // Remember to release the context if we are an entrypoint.
+        IRBuilder<> Builder(&I);
+        Builder.CreateCall(ReleaseCtx, {TheRootContext});
+        ContextWasReleased = true;
+      }
+    }
+  }
+  // FIXME: This would happen if the entrypoint tailcalls. A way to fix would be
+  // to disallow this, (so this then stays as an error), another is to detect
```
Comment (snehasish) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfLowering.cpp:320]: The sentence mentions "disallow this" and then ends with "disallow the tail call". Is this intended?

Disalllowing tailcalls is not trivial unfortunately, while `-fno-optimize-sibling-calls` can stop the compiler from generating tail calls, clang's `musttail` attribute [1] is used in performance critical code. Though I don't think these would be "entrypoints" though. Perhaps we can check this where the entrypoint intrinsics are emitted?

[1] https://clang.llvm.org/docs/AttributeReference.html#musttail

**Review by mtrofin — COMMENTED** [review]
updated

**Review by snehasish — APPROVED** [review]
lgtm

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -14156,10 +14175,10 @@ Syntax:
 Overview:
 """""""""
 
-.. FIXME: detail when it's emitted once the support is added
-
 The '``llvm.instrprof.callsite``' intrinsic should be emitted before a callsite
-that's not to a "fake" callee (like another intrinsic or asm).
+that's not to a "fake" callee (like another intrinsic or asm). It is used by
+contextual profiling and is side-effectful. Its lowering happens in IR, and
```
Comment (snehasish) [inline llvm/docs/LangRef.rst:14180]: "has side-effects"?

**Review by snehasish — APPROVED** [review]
lgtm, a couple of nits on the doc.

**Pull request merged**